### PR TITLE
[HotFix] Introduce retry around storage LoadAsync, SaveAsync, and DeleteAsync

### PR DESCRIPTION
**Pull request by joelverhagen**
These were the methods that were failing intermittently on EA -> Mooncake storage.

### CODE REVIEW

File: src/Catalog/Retry.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            for (int currentAttempt = 0; currentAttempt < maxAttempts; currentAttempt++)
+            {
+                try
+                {
+                    runLogic();
+                    return;
+                }
+                catch (Exception e)
+                {
+                    if (currentAttempt < maxAttempts && shouldRetry(e))
```
Comment (loic-sharma) [inline src/Catalog/Retry.cs]: These could be `when` statements on the `catch`

Comment (joelverhagen) [reply src/Catalog/Retry.cs]: Sure. I just mimicked the existing synchronous implementation but I'm okay with making this change to both.

Comment (loic-sharma) [reply src/Catalog/Retry.cs]: Ah I didn't realize this file was moved. It's fine either ways 😄

File: src/NuGet.Indexing/Retry.cs [hunk 2]
```diff
@@ -1,34 +0,0 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (shishirx34) [inline src/NuGet.Indexing/Retry.cs]: Is it fair to assume this change won't break any other high level dependencies/not used at all? or we are going to treat this as a breaking change?

Comment (joelverhagen) [reply src/NuGet.Indexing/Retry.cs]: `NuGet.Indexing` depends on `NuGet.Services.Metadata.Catalog` so the class is still available to the previous users.

File: src/Catalog/Retry.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                    }
+                    else
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        public static async Task IncrementalAsync(Func<Task> runLogicAsync, Func<Exception, bool> shouldRetry, int maxAttempts, TimeSpan waitIncrement)
```
Comment (dtivel) [inline src/Catalog/Retry.cs]: Is it worth logging a custom metric for attempts > 1?  That way we could easily understand how well this solution works and also how frequently retries are necessary over time.

File: src/Catalog/Retry.cs (hunk 1 above)
Comment (joelverhagen) [reply src/Catalog/Retry.cs]: Done.

File: src/Catalog/Retry.cs (hunk 3 above)
Comment (joelverhagen) [reply src/Catalog/Retry.cs]: Good idea. Done.

**Comment by joelverhagen** [comment]
I am going to hold off on this hotfix until we see another problem running catalog2registration from EA.

**Comment by joelverhagen** [comment]
Closing this PR. Will re-open if needed.

**Pull request closed without merging**